### DISCUSSION TITLE
Does ninja not recognise changes to meson.build file?

### DISCUSSION THREAD

**Discussion by 5wattbulb** [mesonbuild/meson#9755] [Q&A] [answered]
meson.build
```
project('pj', 'c', version : '0.1', default_options : ['default_library=shared'])
library('pj', sources : ['pj.c'], install : true)

```
I execute
```
$ meson setup build
$ ninja -C build

```
All works fine but if I edit meson.build and change the default library to static
```
project('pj', 'c', version : '0.1', default_options : ['default_library=static'])
library('pj', sources : ['pj.c'], install : true)

```
and execute
$ ninja -C build
nothing changes and ninja states 'no work to do'.

The only way I've been able to get it to work is by executing
$ meson setup build --wipe
and then
$meson -C build
However, I'm finding it hard to stomach that you'd have to go through that for every build (e.g. debug, release etc.). I think I'm right in saying that if I execute
$ meson configure build -Ddefault_library=static
then every rebuild will recognise the change but why doesn't ninja just check if the meson.build file has changed and reconfigure automatically for each new meson compile?

**Comment by eli-schwartz** [comment, marked as answer]
Meson does write a build rules that causes ninja to execute the reconfigure command if meson.build is edited.

However, changes to default_options only apply when initializing a build directory from scratch, not when reconfiguring an existing one. Reconfiguring a build directory will always preserve the current value of options.

Changes to meson.build that modify, add, or remove executables, libraries, custom_targets, etc. will absolutely cause a reconfigure that then picks up the new build targets.

**Reply by 5wattbulb** [reply]
Thanks eli. To be honest I'm a bit lost with the terminology and left wondering what meson or ninja's got against default options but at least now I know.

**Reply by eli-schwartz** [reply]
"what meson's got against default options"?

There's nothing wrong with default options... It is just that meson cannot possibly know whether you want to change the option. When you first configured, did you manually specify the same value as the default? Did you leave it unspecified because you wanted the current default anyway? What was the previous default and did it differ from the built-in default?

The simple and reliable answer is to always reconfigure using the same resolved options, and treat the defaults as something that applies at first creation, but not afterward.

This is the same reason that changing the default of a project option in meson_options.txt doesn't change the currently live build.

The current configuration *overrides* the default.

**Comment by 5wattbulb** [comment]
Thanks for your time eli. I'm new to meson (indeed any build software) and what I'm confused about is this. You said in your first post
"Meson does write a build rules that causes ninja to execute the reconfigure command if meson.build is edited."
When I changed the default option from shared to static in my example I edited the meson.build file. If ninja detected I changed meson.build why didn't it reconfigure? How did it know it was just a default option I changed such that it decided to just ignore the edit for some reason?
I suspect I'm looking at this the wrong way but no idea how.

**Reply by eli-schwartz** [reply]
Most build systems, including ninja, use file timestamps to detect out of date targets. Although there are some that use file hashes instead.

If the modification date of `meson.build` changes, ninja will compare timestamps and see that `meson.build` is newer than the build rules (ninja reads a file called `build.ninja`, which is the ninja version of `Makefile`) and it will rerun the configure command. It doesn't try to detect whether edits can be ignored.

On my system, editing `meson.build` and changing the `project()` function's `default_options` causes ninja to do this:

```
[0/1] Regenerating build files.
The Meson build system
Version: 0.60.99
Source dir: /tmp/fooproject
Build dir: /tmp/fooproject/builddir
Build type: native build
Project name: fooproject
Project version: 1.0
C compiler for the host machine: ccache cc (gcc 11.1.0 "cc (GCC) 11.1.0")
C linker for the host machine: cc ld.bfd 2.36.1
C++ compiler for the host machine: ccache c++ (gcc 11.1.0 "c++ (GCC) 11.1.0")
C++ linker for the host machine: c++ ld.bfd 2.36.1
Host machine cpu family: x86_64
Host machine cpu: x86_64
[...]
```

**Reply by 5wattbulb** [reply]
Thanks eli. I just noticed today there's a question "Why are changes to default project options ignored?" in the meson online manual FAQ page.
https://mesonbuild.com/FAQ.html

**Reply by andy5995** [reply]
I just opened a couple of tickets related to this discussion

https://github.com/mesonbuild/meson/issues/9813
https://github.com/mesonbuild/meson/issues/9814
